src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+                                    }
+                                }
+                            } else {
+                                for (final Read record : byPairing.get(false)) {
+                                    record.setDuplicateFragment(true);
+                                    context.output(record);
+                                }
+                            }
+                        }
+                    });
+        }
+
+        /**
+         * Groups reads by keys - keys are tuples of (library, contig, position, orientation).
+         */
+        PCollection<KV<String, Iterable<Read>>> groupReadsByKey(final PCollection<Read> fragments, final PCollectionView<SAMFileHeader> headerPcolView) {
+            final PTransform<PCollection<? extends Read>, PCollection<KV<String, Read>>> makeKeysForFragments =
+                    ParDo.named("make keys for reads")
+                            .withSideInputs(Collections.singletonList(headerPcolView))
+                            .of(new DoFn<Read, KV<String, Read>>() {
+                                @Override
+                                public void processElement(final ProcessContext context) throws Exception {
+                                    final Read record = context.element();
+                                    record.setDuplicateFragment(false);
+                                    final SAMFileHeader h = context.sideInput(headerPcolView);
+                                    final String key = MarkDuplicatesReadsKey.keyForFragment(h, record);
+                                    final KV<String, Read> kv = KV.of(key, record);
+                                    context.output(kv);
+                                }
+                            });
+            return fragments.apply(makeKeysForFragments)
+                    .apply(GroupByKey.<String, Read>create());
+        }
+    }
+
+    /**
+     * How to assign a score to the read in MarkDuplicates (so that we pick the best one to be the non-duplicate).
+     */
+    //Note: copied from htsjdk.samtools.DuplicateScoringStrategy
+    private static int score(final Read record) {
+        if (record == null) {
+            return 0;
+        } else {
+            return record.getAlignedQuality().stream().mapToInt(Integer::intValue).filter(i -> i >= MIN_BASE_QUAL).sum();
+        }
+    }
+
+    /**
+     * Returns the list of intervals from the given sequence dictionary.
+     */
+    private static List<SimpleInterval> getAllIntervalsForReference(final SAMSequenceDictionary sequenceDictionary) {
+    return GenomeLocSortedSet.createSetFromSequenceDictionary(sequenceDictionary)
+            .stream()
+            .map(SimpleInterval::new)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Struct-like class to store information about the paired reads for mark duplicates.
+     */
+    private static final class PairedEnds extends GenericJson{
```

Comment:
Extending GenericJson is fine. An alternative is to define a custom coder (which we'll do once this moves to the Read interface).